tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -8748,6 +8752,14 @@ bool HLSLExternalSource::CanConvert(
           break;
         }
       }
+    } else if (m_sema->getLangOpts().StrictUDTCasting &&
+               (SourceInfo.ShapeKind == AR_TOBJ_COMPOUND ||
+                TargetInfo.ShapeKind == AR_TOBJ_COMPOUND) &&
+               !TargetUnnamed) {
```

Comment:
are there any builtin compound objects that could get implicitly cast?

Reply:
Do you mean unnamed builtin compound objects?  No, I don't think so.  But if there were, I think we would want to preserve this behavior, since otherwise there would be no way to pass something to an intrinsic that used that type.

If that's not what you mean, I'm not sure what you're getting at.